File: api/phase4_roundtrip_audit.py
"""Phase 4 deterministic round-trip audit for exported/imported budget items."""
from __future__ import annotations

from decimal import Decimal, InvalidOperation, ROUND_HALF_UP


def _q(value: object, scale: int) -> Decimal:
    try:
        number = Decimal(str(value if value is not None else "0"))
    except InvalidOperation as exc:
        raise ValueError(f"invalid decimal value: {value}") from exc
    return number.quantize(Decimal(1).scaleb(-scale), rounding=ROUND_HALF_UP)


def audit_roundtrip(source_items: list[dict], imported_items: list[dict], amount_scale: int = 2) -> dict:
    if amount_scale < 0 or amount_scale > 8:
        raise ValueError("amount_scale must be between 0 and 8")
    source = {str(x.get("source_budget_item_id") or x.get("id") or "").strip(): x for x in source_items}
    imported = {str(x.get("source_budget_item_id") or x.get("id") or "").strip(): x for x in imported_items}
    if "" in source or "" in imported:
        raise ValueError("every item must preserve a round-trip lineage id")
    if len(source) != len(source_items) or len(imported) != len(imported_items):
        raise ValueError("duplicate round-trip lineage id")
    missing = sorted(source.keys() - imported.keys())
    added = sorted(imported.keys() - source.keys())
    differences = []
    source_total = Decimal("0")
    imported_total = Decimal("0")
    for item_id in sorted(source.keys() & imported.keys()):
        left, right = source[item_id], imported[item_id]
        left_amount = _q(left.get("amount", _q(left.get("quantity"), 8) * _q(left.get("unit_price"), 8)), amount_scale)
        right_amount = _q(right.get("amount", _q(right.get("quantity"), 8) * _q(right.get("unit_price"), 8)), amount_scale)
        source_total += left_amount
        imported_total += right_amount
        fields = {}
        for field in ("code", "name", "unit"):
            if str(left.get(field, "")).strip() != str(right.get(field, "")).strip():
                fields[field] = {"source": left.get(field), "imported": right.get(field)}
        for field in ("quantity", "unit_price", "amount"):
            if _q(left.get(field), amount_scale) != _q(right.get(field), amount_scale):
                fields[field] = {"source": str(_q(left.get(field), amount_scale)), "imported": str(_q(right.get(field), amount_scale))}
        if fields:
            differences.append({"source_budget_item_id": item_id, "fields": fields})
    source_total = _q(source_total, amount_scale)
    imported_total = _q(imported_total, amount_scale)
    return {
        "consistent": not missing and not added and not differences and source_total == imported_total,
        "amount_scale": amount_scale,
        "source_total": str(source_total),
        "imported_total": str(imported_total),
        "total_difference": str(_q(imported_total - source_total, amount_scale)),
        "missing_lineage_ids": missing,
        "added_lineage_ids": added,
        "item_differences": differences,
    }

File: pcces-go/internal/storage/sqlite/phase4_roundtrip_audit.go
package sqlite

import (
	"fmt"
	"math/big"
	"sort"
	"strings"
)

type RoundTripAuditItem struct {
	SourceBudgetItemID string `json:"source_budget_item_id"`
	ID                 string `json:"id"`
	Code               string `json:"code"`
	Name               string `json:"name"`
	Unit               string `json:"unit"`
	Quantity           string `json:"quantity"`
	UnitPrice          string `json:"unit_price"`
	Amount             string `json:"amount"`
}

type RoundTripItemDifference struct {
	SourceBudgetItemID string                       `json:"source_budget_item_id"`
	Fields             map[string]map[string]string `json:"fields"`
}

type RoundTripAuditResult struct {
	Consistent        bool                      `json:"consistent"`
	SourceTotal       string                    `json:"source_total"`
	ImportedTotal     string                    `json:"imported_total"`
	TotalDifference   string                    `json:"total_difference"`
	MissingLineageIDs []string                  `json:"missing_lineage_ids"`
	AddedLineageIDs   []string                  `json:"added_lineage_ids"`
	ItemDifferences   []RoundTripItemDifference `json:"item_differences"`
}

func auditRat(value string) (*big.Rat, error) {
	if strings.TrimSpace(value) == "" {
		value = "0"
	}
	v, ok := new(big.Rat).SetString(value)
	if !ok {
		return nil, fmt.Errorf("invalid decimal value: %s", value)
	}
	return v, nil
}

func ratText(v *big.Rat) string { return v.FloatString(2) }

func AuditPhase4RoundTrip(sourceItems, importedItems []RoundTripAuditItem) (RoundTripAuditResult, error) {
	result := RoundTripAuditResult{MissingLineageIDs: []string{}, AddedLineageIDs: []string{}, ItemDifferences: []RoundTripItemDifference{}}
	build := func(items []RoundTripAuditItem) (map[string]RoundTripAuditItem, error) {
		out := map[string]RoundTripAuditItem{}
		for _, item := range items {
			id := strings.TrimSpace(item.SourceBudgetItemID)
			if id == "" {
				id = strings.TrimSpace(item.ID)
			}
			if id == "" {
				return nil, fmt.Errorf("every item must preserve a round-trip lineage id")
			}
			if _, exists := out[id]; exists {
				return nil, fmt.Errorf("duplicate round-trip lineage id")
			}
			out[id] = item
		}
		return out, nil
	}
	source, err := build(sourceItems)
	if err != nil {
		return result, err
	}
	imported, err := build(importedItems)
	if err != nil {
		return result, err
	}
	for id := range source {
		if _, ok := imported[id]; !ok {
			result.MissingLineageIDs = append(result.MissingLineageIDs, id)
		}
	}
	for id := range imported {
		if _, ok := source[id]; !ok {
			result.AddedLineageIDs = append(result.AddedLineageIDs, id)
		}
	}
	sort.Strings(result.MissingLineageIDs)
	sort.Strings(result.AddedLineageIDs)
	sourceTotal, importedTotal := new(big.Rat), new(big.Rat)
	for id, left := range source {
		right, ok := imported[id]
		if !ok {
			continue
		}
		la, err := auditRat(left.Amount)
		if err != nil {
			return result, err
		}
		ra, err := auditRat(right.Amount)
		if err != nil {
			return result, err
		}
		sourceTotal.Add(sourceTotal, la)
		importedTotal.Add(importedTotal, ra)
		fields := map[string]map[string]string{}
		for name, pair := range map[string][2]string{"code": {left.Code, right.Code}, "name": {left.Name, right.Name}, "unit": {left.Unit, right.Unit}, "quantity": {left.Quantity, right.Quantity}, "unit_price": {left.UnitPrice, right.UnitPrice}, "amount": {left.Amount, right.Amount}} {
			if strings.TrimSpace(pair[0]) != strings.TrimSpace(pair[1]) {
				fields[name] = map[string]string{"source": pair[0], "imported": pair[1]}
			}
		}
		if len(fields) > 0 {
			result.ItemDifferences = append(result.ItemDifferences, RoundTripItemDifference{SourceBudgetItemID: id, Fields: fields})
		}
	}
	result.SourceTotal = ratText(sourceTotal)
	result.ImportedTotal = ratText(importedTotal)
	result.TotalDifference = ratText(new(big.Rat).Sub(importedTotal, sourceTotal))
	result.Consistent = len(result.MissingLineageIDs) == 0 && len(result.AddedLineageIDs) == 0 && len(result.ItemDifferences) == 0 && sourceTotal.Cmp(importedTotal) == 0
	return result, nil
}
